### DISCUSSION TITLE
Alternative Shield Improvements

### DISCUSSION THREAD

**Discussion by xX-Dillinger-Xx** [endless-sky/endless-sky#9811] [Balance, Mechanics, and Interactions]
I was recently involved in the discussion in [Amazinite's issue](https://github.com/endless-sky/endless-sky/issues/9802) about HP and DPS. I'm also aware that Quantumshark has a [PR](https://github.com/endless-sky/endless-sky/pull/9771) testing some ideas he has involving shield delays. I decided to start another related topic here instead of flooding their topics.
 
I'm offering an alternative approach as to how shields work and it may address some of the complaints about small ships and survivability. I don't have the skill set to do a PR of my own so I'll stick to discussions. If someone want's to try this idea in a PR, feel free. 

Simply put, shield generators should have 3 attributes. Size, Speed and Capacity ( as well as the existing ones like energy, price ect. )
size - we already have
speed - we already have
capacity - is the max amount of shields it can maintain. If the generator's capacity exceeds that of the ship, we have 2 options. 1: add the difference to a ships shields OR 2: cap the capacity to the max capacity of the ships shields. I'm thinking adding makes sense. 

- small generators should be faster but have lower capacity. 
- larger generators should be slower with a higher capacity.
I think of a ships shield as a matrix that the generator forms the shields over. A shield generator can only maintain a certain size shield matrix. The capacity the generator has determines it's speed. If your ship requires more capacity then your current generator has, you need to upgrade. 
 
**Examples:** (using numbers for explanation purposes only. The real numbers are to be determined for each race and tier.) 

Shield Gen. A 
- outfit space 20
- regen speed  100/sec
- max capacity 1000 

Shield Gen. B
- outfit space 30
- regen speed  75/sec
- max capacity  2000 

Shield Gen. C 
- outfit space 40
- regen speed 50/sec
- max capacity 3000 

**Scenarios**
Assume we have a ship with 1500 shields.  
A: ( over capacity option 1 ) We have Shield Gen. B installed with a max capacity of 2000, we start with 2000 ( gens. max  - so we gained 500 from the generator), we take damage and **after regen** we 2000. This is my preferred option. 

B: ( over capacity option 2 ) We have Shield Gen. B installed with a max capacity of 2000, we start with 1500 (ships max), we take damage and  **after regen** we have 1500. 

C: We have Shield Gen. A installed with a max capacity of 1000, we start with 1500 ( ships max cap. ), we take damage and  **after regen** we have 1000 (gen's max cap. ) until you land. 

Scenario A:  you get a boost to shields but regens slower
Scenario B:  you don't get any advantage from an over capacity generator but you can reach ships max, regen is still slower
Scenario C: you get faster regen but loose a little bit of shields, after an attack. This would encourage you to upgrade.

A and C balance each other. Something that might help this make more sense within the game universe is that the current "ships shield stat" would now represent the size of the matrix that the ship has.  Technically, the number doesn't need to change, just how we think of it's meaning.   

This is just a rough idea, and I can't support my own logic for some of it at the moment, but I think you can get the idea. My thinking is smaller ships need help and larger ships already have huge shields anyway. I have additional thoughts on shields and regen but will save it for later, if anybody is interested.

**Comment by Hecter94** [comment]
It's an interesting idea; one thing to note is that for this:
> capacity - is the max amount of shields it can maintain. If the generator's capacity exceeds that of the ship, we have 2 options. 1: add the difference to a ships shields OR 2: cap the capacity to the max capacity of the ships shields. I'm thinking adding makes sense.

We'd need to go with Option 2. Increasing ship HP pools is something we avoid doing as it becomes very difficult to balance appropriately and takes away a lot of the uniqueness of each ship. 
(Imagine the Emerald Sword but stuffed full of shield outfits to let it brawl with T3 warships)

One question I do have is how exactly these stack. 
If I have two shield generators, each with a capacity of 3000, installed on a ship with 5000 shields, will my ship regen up to 3000? Or 5000? Assuming we are capping the max shields to the ship's base shields, like I mentioned. 

Something to note is that the game engine only allows outfit stats to take the max out of all installed outfits or sum them all together into one big lump sum. 
So really, the only two options are:
"The capacity of every shield generator installed is added together."
or
"We take the highest capacity out of every installed shield generator and use that." 

Both have issues; imagine installing a single generator with a large capacity and slow regen and pairing it with another generator with a low capacity but high regen. 
No matter how you add those stats together, the result is high capacity with high regen, which is not ideal. 

You could start limiting players to one shield generator per ship, but those sorts of arbitrary limits are another thing we try to avoid. There are very few outfits in the game that you can only install a limited number of, and I'm not sure making shields one of them would be an improvement.

**Reply by xX-Dillinger-Xx** [reply]
> (Imagine the Emerald Sword but stuffed full of shield outfits to let it brawl with T3 warships)

 
 That's no different then using an Arfecta. LOL I wasn't thinking to make that possible anyway. I too want to preserve uniqueness.

> One question I do have is how exactly these stack.

I don't really have an answer for that. It was just a rough concept idea. A lot would be need to determined by people other then me. I didn't see much point in wasting time figuring out every last detail if the devs aren't interested. Which, BTW looks to be the case. 

> Both have issues; imagine installing a single generator with a large capacity and slow regen and pairing it with another generator with a low capacity but high regen.
No matter how you add those stats together, **_the result is high capacity with high regen_**, which is not ideal.

Not really. Lets use these 2 for a second and a ship with 4000 shields. 
- speed of 100/sec and 1000 cap.
- speed of 50/sec and  3000 cap.

You take a lot of damage. What you get is  the first 1000 regen at 100/sec and the rest at 50/sec. (Maybe, this might work with Quantumshark's shield delay project.) However, what would stop someone from just stacking high speed, low cap. units? I don't know. Maybe, forget about the Generators capacity all together and just base it's speed of the ships shields. The more base shields a ship has the slower the regen speed, within reason of course.

**Comment by alexrovw** [comment]
If we want something like Scenario C, where we are limiting the amount of shield that a ship can recover without landing, then Azure's PR implements this. https://github.com/endless-sky/endless-sky/pull/9690

On a higher level, I think mucking with the amount of shield hp that a ship can support is messy business. Shield hp should remain a feature of each individual ship, since each ship is balanced around that number. You should not be able to add to the amount of max shield that a ship has.

**Reply by xX-Dillinger-Xx** [reply]
I wasn't suggesting to change the balance. The current shield value would remain the same.  Just small advantages or disadvantages. Each ship would still be unique with it's own base shield amounts.

**Comment by Saugia** [comment]
I do want to leave a note that one thing we're not looking as a possibility is capacity for shields (and at that rate, armor). Over the years for Endless Sky there have been many discussions on having outfits that add shield and hull values, but we stand firm on the statement that we do not want outfits to add such as it would be a balancing nightmare amongst other concerns.

I'm hoping to have a documentation of our vision of the game as well as what aspects we're interested in and what has been brought up often and been either accepted or rejected as a sort of Q&A guide because not everyone will see the past discussions and thoughts on everything over periods of years.

~~I see the others also mentioned that, my bad. But yeah that's the deal with that.~~ (I will add there may be a *tiny* chance we add very exclusive or rare outfits that do that, but nothing as an overall general design.)

**Reply by alexrovw** [reply]
Just wanted to add the caveat that this is for Vanilla, I presume. We already have `"shield multiplier"` and `"hull multiplier"` attributes, which plugin authors can take advantage of to create outfits that increase max shields/hull.

**Reply by xX-Dillinger-Xx** [reply]
@Saugia Documentation what be a huge help
@alexrovw Yes I was thinking of Vanilla. I'm not a big fan of plugins anyway. I haven't really checked out any ES ones. Typically, plugins tend to venture too far from what the vision of the game was intended to be. Not, to mention poor quality. I'm not suggesting that's the case here. I've only check out one and when I use anything from it I feel like I'm cheating. I know it's stupid, that's just me.

**Comment by xX-Dillinger-Xx** [comment]
Like I said, it was just a rough idea designed to start a discussion. I hoped others would offer some tweaks, so thanks for that. I wonder if any of this could be merged into what Quantumshark is working on.  The main objective of the idea is to allow smaller ships to have faster regen then larger ships without allowing those same faster generators to be used on larger ships. I'm just not sure how we get there. Maybe adopt the modular approach that's being used with Bunrodean outfits.

I respect the decision to not allow outfits that rise HP , but I'm not sure how that would create a balance issue. Ships are already extremely customizable, which changes balance. How does a little extra shields or hull make that much difference when it only applies to the players ships? All the NPC ships would remain as they where designed.
IMO balance really goes out the window as soon as the player caps or buys a T2+ ship and brings it back to human space. Let's not mention what happens when you cap an Arfecta. It's game over for every NPC after that. in fact I have an Arfecta and hardly use it. For me, no challenge, no fun. Just to be clear, I wasn't thinking to add 10000 shields to a Wasp LOL. BTW I'm not arguing just pointing some stuff out.

**Reply by Quantumshark** [reply]
> The main objective of the idea is to allow smaller ships to have faster regen then larger ships without allowing those same faster generators to be used on larger ships.

I intend to accomplish this using the mechanics changes from #9771. Because the delay time stacks additively, stacking smaller generators on larger ships is undesirable, and larger ships are less able to disengage and/or evade fire too.
![image](https://github.com/endless-sky/endless-sky/assets/18634983/7fcd5b53-e1ce-420b-8e95-487d5a4334f4)
I've worked out some numbers here that give larger generators better continuous generation, but give smaller generators the better regen/ton after disengaging, which should give a huge edge to smaller ships relative to what we have now, especially combined with the speed changes in #9616. I think between those two changes, we'll have changed enough to solve the large ship / small ship balance issues, but at the very least we ought to give the combination of those two adequate testing before we need to try coming up with further ideas.

**Reply by Saugia** [reply]
Indeed. One of the more discussed options is what you have mentioned, and that's having smaller ships have higher regen rates while larger ships have slower and/or delayed regen rates, so that is definitely on the table as a possibility.

Having an outfit give or take shield and hull values is as mentioned a balance nightmare - it's very hard to balance the game around such items as you can effectively increase the hp pool of anything you want when weapons and ship health are balanced around static values, not around the idea if a player bought said items.

**Reply by xX-Dillinger-Xx** [reply]
@Quantumshark I really like your idea and it looks like it will accomplish what I had imagined. So feel free to incorporate anything from my idea if you want.  I won't be doing anything with it personally. If you don't that's fine too. As I have mentioned before, I can come up with ideas, I just can't implement them. My ideas are to make people think of alternatives. I spent my whole career using the motto, "There's always another way, I just haven't thought of it yet"
